Try adding layout classnames to inner block wrapper

tellthemachines opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

As a prequel to #43140 we need a way to attach layout-related classnames to the inner block wrapper, instead of the outer (they are not always the same).

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Does some searching and matching to work out where the inner wrapper classnames are, and attach layout classes to those.


## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

I added layout to the Cover block so we can test this is working. It doesn't have to stay as part of this PR; we can just focus on the mechanism for adding classes and then deal with block-specific stuff later. But for now this can be tested by adding a Cover block with some content and testing its layout controls. 

1. Add Cover block with a couple blocks inside;
2. Change controls in "Layout" section of block sidebar;
3. Check that layout classnames are attached to inner block wrapper on both editor and front end;
4. Check that other blocks with layout (e.g. Group, Buttons, Columns...) are still working correctly.

## Screenshots or screencast <!-- if applicable -->

tellthemachines committed:
Try adding layout support to Cover block.

tellthemachines committed:
Try adding layout classnames to inner block wrapper.

andrewserong commented:
Thanks for digging into this one! It looks like the WP_HTML_Tag_Processor class has been merged as of https://github.com/WordPress/gutenberg/pull/42485 — would it be worth trying that out for this case, to avoid adding an additional regex? That class isn't slated to land until 6.2, so assuming that this feature is also for 6.2, I was wondering if it might be a good opportunity to see if the class holds up for this use case 🤔

tellthemachines commented:
> the WP_HTML_Tag_Processor class has been merged as of https://github.com/WordPress/gutenberg/pull/42485 — would it be worth trying that out for this case, to avoid adding an additional regex?

I'll look into it! We should be able to use it for the part where we add the layout classes. I doubt if it'll work for identifying the inner wrapper though, as we're going by its position in the `$inner_content` array. We have no way of knowing what classnames or tags it might have, which means there's nothing for `WP_HTML_Tag_Processor` to work with there.

andrewserong commented:
> We have no way of knowing what classnames or tags it might have, which means there's nothing for WP_HTML_Tag_Processor to work with there.

Ah, good point, yes, I forgot about the problem of flagging where the wrapper should ideally be!

tellthemachines committed:
Make it work in the editor

tellthemachines commented:
Update: got it working on the editor, but there's some duplication of classnames if the inner wrapper isn't the same as the outer one. I found we can't rely on InnerBlocks to add the classnames for all container blocks, because some don't use it (e.g. Gallery) and others only use it in certain views (e.g. PostContent). But I haven't found any reliable way to remove classes from the outer wrapper if they are applied to the inner already. Ideas appreciated!

Still todo: look into using `WP_HTML_Tag_Processor`.

tellthemachines committed:
Use WP_HTML_Tag_Processor to add layout classnames

tellthemachines commented on packages/block-editor/src/components/block-list/block.js:
```diff
@@ -231,7 +233,8 @@ function BlockListBlock( {
 			dataAlign && themeSupportsLayout && `align${ dataAlign }`,
-			className
+			className,
+			layoutClassNames
```
We need to add these here for blocks that don't use `InnerBlocks` to render their children (the only examples I could find of this in Core are Gallery and PostContent when viewed in non-editable mode in the site editor, but there may be more). Unfortunately that means that if the block's inner wrapper is different from its outer, the layout classnames will be duplicated across both wrappers in the editor. There don't seem to be any consequences to this, as the blocks that currently have multiple wrappers and layout (e.g. Navigation) are already  handling the styles being applied to their outer wrapper. But it would be nice to fix if anyone has ideas!

andrewserong replied:
Because the constrained layout and root padding features have rules that target the direct child of the layout classname, I think we might need to come up with a way where the layout classes are only applied to the deepest node (that is, the innerBlocks wrapper, or the outer wrapper if no inner wrapper is in use). I'm not quite sure how we ensure that, though! 😅

tellthemachines replied:
The only alternative I can see here is to be deliberate about layout _not_ supporting out of the box blocks that don't use InnerBlocks, and add some custom logic to make things work in Gallery and Post Content. I'll look into that!

tellthemachines committed:
Use WP_HTML_Tag_Processor to identify inner wrapper classnames

dmsnell commented on lib/block-supports/layout.php:
```diff
@@ -291,6 +291,28 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
+
+	while ( true === $inner_content_chunk->next_tag() ) {
+		$inner_wrapper_classnames = $inner_content_chunk->get_attribute( 'class' );
+	}
```
a couple thoughts on this function:

we're passing in inner blocks but then trying to get the first one. the additional list-handling here might not be helping the clarity of its purpose. if we passed in _a block_ instead of a list of blocks, or instead of _inner content_, then it could focus on a more specific purpose.

```php
function get_classes_from_last_tag( $html ) {
	$tags         = new WP_HTML_Tag_Processor( $html );
	$last_classes = '';

	while ( $tags->next_tag() ) {
		$last_classes = $tags->get_attribute( 'class' ) || '';
	}

	return $last_classes;
}
```

Secondly, while I think the code as-is is fine, we don't need the `true ===` in the `while` condition because as a boolean value that's already implicit. No problem keeping it in there, but this isn't like functions which return a _truthy_ value; it actually returns `true` or `false`.

Third, `get_attribute( 'class' )` might return `null` if the last tag doesn't have a `class` attribute. The return type in the docblock comment says this function returns a `string` but unless we add a fallback it will also return `null` from time to time.

Fourth, a block's `$inner_content` is not guaranteed to contain only strings. It may contain `null` values where inner blocks belong. Another benefit to focusing this function on HTML processing and letting the calling code extract the HTML it wants to scan is not having to deal with this confusion here.

```php
$inner_content_classnames = isset( $block['innerContent'][0] ) && 'string' === typeof $block['innerContent'][0]
	? get_classes_from_last_tag( $block['innerContent'][0] )
	: '';
```

@adamziel I wonder; if we exposed the offset/position in the stream we could allow people to add their own "rewind" tracking and other more complicated parsing. would that be worth the risk it exposes?

```php
$last_tag_at = null;
while ( $tags->next_tag() ) {
	$last_tag_at = $tags->current_position();
}

if ( null === $last_tag_at ) {
	return;
}

$tags->seek( $last_tag_at );
return $tags->get_attribute( 'class' ) || '';
```

this could expose a lot of functionality it seems people want to do with this, as long as we don't expose the internal pointer, such that the offset points to an old version of the document, before it was changed by `set_attribute()` calls… 🤔

tellthemachines replied:
Ooh, that's great feedback, thanks! Yes, the function is a lot neater in your example 😄

adamziel replied:
> @adamziel I wonder; if we exposed the offset/position in the stream we could allow people to add their own "rewind" tracking and other more complicated parsing. would that be worth the risk it exposes?
 
Hm hm hm. This worries me in the same way as exposing `get_classes()` [worried you in the other thread](https://github.com/WordPress/gutenberg/pull/44600#discussion_r997552055). The cursor is an implementation detail and I'd rather not have people mess around with it – it would make their core fragile and severely constraint the future updates to the tag processor. For example, `seek` opens the following can of worms:

```php
$tags->next_tag();
$tag_1_at = $tags->current_position();

$tags->next_tag();
$tag_2_at = $tags->current_position();

$tags->seek( $tag_1_at );
$tags->set_attribute( 'class', 'bold' );

// $tag_2_at no longer points where we expect it to
$tags->seek( $tag_2_at );
$tags->set_attribute( 'class', 'bold' );
```

I'd rather keep an eye out for the use-cases that come up and see if we can address most of them with case-specific APIs. For example, this loop could benefit from `has_next_tag()`:

```php
while($tags->has_next_tag()) {
    $tags->next_tag();
}
return $tags->get_attribute( 'class' );
```

dmsnell commented on lib/block-supports/layout.php:
```diff
@@ -412,16 +433,16 @@ function gutenberg_render_layout_support_flag( $block_content, $block ) {
+		$content->add_class( esc_attr( implode( ' ', $class_names ) ) );
+	} else {
+		$content->next_tag();
+		$content->add_class( esc_attr( implode( ' ', $class_names ) ) );
```
there's no need or benefit to running these through `esc_attr()` since `add_class()` already does this (by means of its call to `set_attribute()`)

tellthemachines replied:
Thanks, that's good to know!

dmsnell commented on lib/block-supports/layout.php:
```diff
@@ -412,16 +433,16 @@ function gutenberg_render_layout_support_flag( $block_content, $block ) {
+
+	$content = new WP_HTML_Tag_Processor( $block_content );
+	if ( $inner_content_classnames ) {
+		$content->next_tag( array( 'class_name' => $inner_content_classnames ) );
```
this does what you want it to do, but it surprised me that it did, because we intended to search for individual class names and not the entire list of classes. it would be good to be aware that if we were to chop out one of the classes this would fail because the classes in `$inner_content_classnames` would no longer be an identical string of the same exact length as the `class` attribute in the searched HTML

tellthemachines replied:
I think it's safe to take that risk here because we haven't changed anything in `$block_content` up until this point, and we don't have a better way of targeting the block's inner wrapper. Potentially, if we had a way to know ahead of time which tag is the last, we could add a classname to it instead of storing the existing ones, and then remove it after this step.

adamziel replied:
@tellthemachines you could filter the tag openers manually:

```php
$sought_class_names = $inner_content_classnames;
while( $content->next_tag() ) {
    $tag_class_names = preg_split('/\s+/', $content->get_attribute( 'class' ));
    if ( /* the two overlap */ ) {
      // ...
    }
}
```

I dislike the `preg_split`, though. @dmsnell We already have `get_attribute()` in the processor, any reason not to add `get_classes()`?

dmsnell replied:
@adamziel my spidey-sense suggests it wouldn't be that much of an added value above `get_attribute()`. perhaps `has_class()` though would be.

what I _don't_ want people to do is what we have in your code example. we provide semantic operations to add and remove CSS classes. if we're encouraging people to extract the raw value(s) and perform their own twiddling then we're just pushing them away from a robust framework and towards multiple ad-hoc implementations of the same thing.

it highlights the _elephant in the room_, which are class names with HTML entities. we may have to come back and revisit those; maybe we could find a way to detect if we think we have an entity (a character is `&`) and then rewind the class comparison with a cloned and decoded value.

point being is that I think still it's more ideal to focus on intended behaviors instead of internal mechanics. _why_ do we want `get_classes()` when we can already query by tags containing a class name and `get_attribute( 'class' )`

adamziel replied:
Such a good point @dmsnell, let's explore adding `has_class`. I added a note to the overview issue's description so we don't lose track of it.

getdave replied:
Here's a good example of where this would be useful

https://github.com/WordPress/gutenberg/pull/40778/commits/baef2619d79a1fcdd54c1b0d4dfdc64d481a4d76

I was assuming it already had such a method!

adamziel replied:
+1 for an idiomatic `has_class` as the one-line implementation we're likely going to see more of does not account for nuances like `class="not-current-menu-item"`:

```php
if ( strpos( $inner_blocks_html, 'current-menu-item' ) ) {
     // ...
}
```

cc @dmsnell

adamziel replied:
@dmsnell @aristath @getdave here's a PR that adds `has_class` https://github.com/WordPress/gutenberg/pull/46232

dmsnell replied:
@getdave what is the good example in that `bugfix` link? what was the bug? I'm trying to understand how `has_class()` would be a help there.

taking a guess, it looks like the goal is to only seek out `wp-block-navigation-item__content` if there is a `current-menu-item` in the document, and I'm not sure where `has_class()` comes in there.

if we want to know if any inner block HTML contains the class, the tag processor already includes this functionality.

```php
$inner_tags = new WP_HTML_Tag_Processor( $inner_blocks_html );
$has_current_menu_item = $inner_tags->next_tag( [ 'class_name' => 'current-menu-item' ] );
```

Secondly, I don't know that the `strpos()` check is all that bad. If we're looking to make sure that we don't introduce needless slow-down, then doing a quick string search over the inner HTML _for the possibility of finding the class name_ might be a good way to keep lots of requests fast while only resorting to proper HTML parsing when we have reason to believe the class might exist within.

adamziel replied:
@dmsnell Using `next_tag` like that is very clever! And now that I think about it, it's also fully within what it was invented for, it just didn't jump out to me in the first place. 

> I don't know that the `strpos()` check is all that bad.

Now that I think about it, the `class_name` (and `tag_name`) matchers could run that `strpos` check before trying to match tags. It would work well except for matching html entities.

dmsnell replied:
@adamziel apart from the HTML entities, which we currently don't match, the `strpos` speedup might warrant more justification before putting it in the tag processor.

on the one hand, `strpos()` is going to be faster if the wanted class isn't in plaintext in the document, but on the other hand, it scans up to the end of the document every time, and if we expect to find another tag "soon" after where the current pointer is, there's no need to repeatedly iterate through the document.

in terms of the general case I'd really want measured data showing that the additional complexity is worth it. my guess is that sometimes it would be faster and sometimes it would be slower, but in all cases it would add additional complexity.

dmsnell replied:
https://github.com/WordPress/wordpress-develop/pull/5096

andrewserong commented on packages/block-editor/src/components/inner-blocks/index.js:
```diff
@@ -200,11 +200,15 @@ export function useInnerBlocksProps( props = {}, options = {} ) {
 		innerBlocksProps.value && innerBlocksProps.onChange
 			? ControlledInnerBlocks
 			: UncontrolledInnerBlocks;
+	//Dedupe layout classes
+	const allTheClassNames = `${ props.className } ${ layoutClassNames }`;
+	const classNameSet = new Set( allTheClassNames.split( ' ' ) );
+	const dedupedClassNames = Array.from( classNameSet ).join( ' ' );
```
Rather than de-duping after the fact, is there a way to ensure that layout classnames are only ever added once? I'm not quite sure how we'd do this, but ideally we'd only be outputting `layoutClassNames` in one place, with preference for the inner blocks wrapper where present. Since `useInnerBlocksProps` occurs after `useBlockProps`, I can't think of a clever way to do that, though 🤔

tellthemachines replied:
This is related to the comment above: the duplication here happens because we're adding layout classnames to BlockListBlock. If we don't do that, this won't be a problem.

>  I can't think of a clever way to do that, though

Can you think of any non-clever ways? 😅

andrewserong replied:
The only non-clever thing I could come up with was to avoid `useBlockProps` and `useInnerBlocksProps` altogether, and create a separate hook that grabs `layoutClassNames` from the block edit context (touched on [in this comment](https://github.com/WordPress/gutenberg/pull/44600#issuecomment-1276860156)).

It would mean saying, something along the lines of, "hey, the other block supports happen automatically, but if you want to use layout, you've got to use the layout hook to grab the stuff you need, and then output it where it should go" 😅 . If it worked, I'm wondering if something like the following pseudo-code might get us in somewhere in that direction:

```js
// layout classnames hook
function useLayoutClassNames( props ) {
  const { layoutClassNames = '' } = useBlockEditContext();
  let mergedClassnames = '';

  // insert code here to merge props.className with layoutClassNames

  return mergedClassnames;
}

// a block's edit component
function MyBlockEdit( props ) {
  const blockProps = useBlockProps();
  const innerBlocksProps = useInnerBlocksProps( blockProps );  

  // The following gets the layout class names and merges them with the classnames from the provided props
  const layoutClassNames = useLayoutClassNames( innerBlocksProps );

  return <div { ...blockProps }>
    <div { ...innerBlocksProps } className={ layoutClassNames } />
  </div>
```

Not sure if something like that's at all viable, but it means that we'd have a hook for getting / merging layout classnames, and the block's Edit component would then be responsible for where to put them. It would mean that becomes part of the API of the layout support for blocks that want to use it, so probably not something that's all the desirable, because it wouldn't "just work" in the way that we typically expect block supports to work.

But that's all I could think of for now!

andrewserong reviewed (commented):
This is looking very cool @tellthemachines, thanks for digging in! It's going to add a lot of nice flexibility to the cover block 👍

The server-rendering appeared to be working pretty well in my testing so far, but I ran into an issue with the edit view, though, because of the duplicate output of layout classnames. The constrained and root padding aware rules target the direct children of the layout classnames, and so assume that they're only going to be added on the direct wrapper of child blocks. This isn't so noticeable with the flex layout, but if we set the cover block to constrained, then in the editor view, the cover block's `span` winds up being constrained unexpectedly since the `wp-container-$id` class is added to the outer wrapper as well as the inner wrapper. On the site frontend, where the classnames are only injected on the inner wrapper, the issue isn't present:

| Editor (layout classnames are on both wrappers) | Site frontend (layout classnames are only on inner wrapper |
| --- | --- |
| <img width="1273" alt="image" src="https://user-images.githubusercontent.com/14988353/195466753-c4e0fb42-7bb0-48f0-9314-154299da858e.png"> | <img width="1268" alt="image" src="https://user-images.githubusercontent.com/14988353/195466855-f469072c-9add-4ee4-bd71-95ade4417ecb.png"> |

I've added a couple of comments, but from a quick look, I can't quite work out an alternative — from your comment about the the `Gallery` and `PostContent` blocks, it sounds like we can't rely on only outputting `layoutClassNames` in `useInnerBlocksProps`?

One of the things I was wondering about, is if there are particular blocks that have edge cases — is that something that can be solved for those individual blocks (the Post Content block is already quite an unusual one), or is it better to see if we can ensure the API for this catches all cases?

andrewserong commented:
Just thinking out-loud, but would another potential alternative for the edit view side of things be to create a separate `useLayoutClassNames` hook that blocks use, and can then pass those classnames directly to where they'd like the layout classnames to be output? Similar to `useBlockProps` and `useInnerBlocksProps`, `useLayoutClassNames` would retrieve the `layoutClassNames` from `useBlockEditContext();`, so the edit component would then be able to place the classnames where they like. It could help bypass the problem of the other hooks needing to know whether or not they should be outputting the layout classnames?

The downside is that layout classnames wouldn't be automatically handled by `useBlockProps` or `useInnerBlocksProps`, but a potential positive is that blocks' edit components could be explicit about where to place the output.

That might just wind up creating more boilerplate, though, so possibly not be a practical option 🤔

ramonjd commented:
Testing this as well. Looking great so far! This is what I'm seeing:


https://user-images.githubusercontent.com/6458278/195477226-f2663b3d-a850-4cb9-a41a-ee7476669554.mp4

> This isn't so noticeable with the flex layout, but if we set the cover block to constrained, then in the editor view, the cover block's span winds up being constrained unexpectedly since the wp-container-$id class is added to the outer wrapper as well as the inner wrapper.

I noticed this as well. This is a screenshot of the frontend HTML, where the unique layout class that controls the max-width is not applied to all the cover block's children:

<img width="998" alt="Screen Shot 2022-10-13 at 12 08 30 pm" src="https://user-images.githubusercontent.com/6458278/195477307-ba4d4082-c841-4f67-8eaf-c46821b9a9cf.png">

Does it sound too dodgy to move the unique layout class to the block container in `cover/index.php`?

We're already doing some [magic with the inner container](https://github.com/WordPress/gutenberg/blob/303c6ad0ccd56e19762d364ccad692c7a728163a/packages/block-library/src/cover/index.php#L39). I was testing it manually in the browser and it _seemed_ to be okay. 😅 

Also, if you think it's appropriate in this PR, it looks like there's another opportunity to use the `WP_HTML_Tag_Processor` class in [cover/index.php](https://github.com/WordPress/gutenberg/blob/303c6ad0ccd56e19762d364ccad692c7a728163a/packages/block-library/src/cover/index.php#L59)

Something like: 

```php
$p = new WP_HTML_Tag_Processor( $content );
$p->next_tag();
$p->set_attribute( 'style', $styles );
```

andrewserong commented:
> This is a screenshot of the frontend HTML, where the unique layout class that controls the max-width is not applied to all the cover block's children:

For that one, I thought the server-rendered view looked a bit better, as the `span` and `<img>` elements would remain full-width, but only the children of the inner container would get `contentSize` etc, applied. Whereas in the edit view, it's affecting both levels (container > img/span, as well as children of the inner wrapper) 🤔

ramonjd commented:
> I thought the server-rendered view looked a bit better, as the span and <img> elements would remain full-width, but only the children of the inner container would get contentSize etc, applied. Whereas in the edit view, it's affecting both levels (container > img/span, as well as children of the inner wrapper) 🤔


Oh good point. My only reaction was that the editor and frontend weren't same when adding a constrained content size, so either way... 😄

tellthemachines commented:
Thanks everyone for all the feedback and testing!

> One of the things I was wondering about, is if there are particular blocks that have edge cases — is that something that can be solved for those individual blocks (the Post Content block is already quite an unusual one), or is it better to see if we can ensure the API for this catches all cases?

I'm starting to think it's better to optimise for blocks that use InnerBlocks, which is the canonical way to handle block nesting, and for edge case situations that require custom nesting logic, it won't be too much of a burden to handle layout manually as well. 

> Just thinking out-loud, but would another potential alternative for the edit view side of things be to create a separate useLayoutClassNames hook that blocks use, and can then pass those classnames directly to where they'd like the layout classnames to be output?

This could be a useful tool for handling the edge cases! but I wouldn't make it the default solution. At least not if we want to move towards a Layout support that Just Works for the majority of use cases 😅

tellthemachines commented:
> Also, if you think it's appropriate in this PR, it looks like there's another opportunity to use the WP_HTML_Tag_Processor class in [cover/index.php](https://github.com/WordPress/gutenberg/blob/303c6ad0ccd56e19762d364ccad692c7a728163a/packages/block-library/src/cover/index.php#L59)

I added layout to Cover in this PR for testing purposes only, and plan to remove it before merging. Mainly because there's a bit of discussion going in #43140 around the best way to handle multiple layout types plus the matrix alignment tool. That sounds like a good candidate for a forthcoming Cover-specific PR though!

andrewserong commented:
> This could be a useful tool for handling the edge cases! but I wouldn't make it the default solution. At least not if we want to move towards a Layout support that Just Works for the majority of use cases

Ah, we said much the same thing at the same time! 😄

Yes, I agree, it'd be good to see how far we can get with the "just works" approach, and what the edge-casey workarounds look like for the other blocks 👍. In terms of API, perhaps we can think of it as the part that "just works", and then additional tools / hooks, etc, for edge-casey blocks to use?

tellthemachines committed:
Improvements to inner content classname logic

tellthemachines committed:
Only add layout classnames to inner blocks

tellthemachines committed:
Manually add layout classnames to Gallery and Post Content edit

tellthemachines commented:
Update: turns out the edge case workarounds were pretty easy, due to `layoutClassNames` already being passed to BlockEdit.

andrewserong commented on lib/block-supports/layout.php:
```diff
@@ -412,16 +428,16 @@ function gutenberg_render_layout_support_flag( $block_content, $block ) {
-		$block_content,
-		1
-	);
+	$inner_content_classnames = isset( $block['innerContent'][0] ) && 'string' === gettype( $block['innerContent'][0] ) ? gutenberg_get_classnames_from_last_tag( $block['innerContent'][0] ) : '';
```
Is it worth adding a comment to explain what the `0` index of `innerContent` is, since it's sort of internal to the structure of the block object? E.g. something along the lines of:

```
// Attempt to extract classnames from last tag in block markup directly preceding first inner blocks marker.
// This identifies the first, deepest nested inner blocks wrapper.
```

Please correct me, though, if I've misunderstood how this is working, I just [looked up the block parser class](https://github.com/WordPress/wordpress-develop/blob/3b63a75108b7f6d5e9112c556285ab1bc80ddbcf/src/wp-includes/class-wp-block-parser.php#L62), so I _think_ that's what it's doing, but could very well be wrong 🙂

From the comment in that class, it sounds like there's the potential for blocks to have another string between `null` markers (on [this line](https://github.com/WordPress/wordpress-develop/blob/3b63a75108b7f6d5e9112c556285ab1bc80ddbcf/src/wp-includes/class-wp-block-parser.php#L62) of the comments). I've only ever encountered an opening string, multiple `null` values for each of the inner blocks, and then a closing string, so I wonder if there are any blocks that do that in practice? If so, I reckon we could look at that if and when we encounter it and then come back to the logic in this function to re-examine it, rather than worry about it now.

For now, looking at the last tag in the opening string sounds good to me!

tellthemachines replied:
Yes, that's exactly how it works! Good idea to add a comment, as it's not immediately obvious.

andrewserong commented on lib/block-supports/layout.php:
```diff
@@ -397,7 +413,7 @@ function gutenberg_render_layout_support_flag( $block_content, $block ) {
 
 		$style = gutenberg_get_layout_style(
-			".$block_classname.$container_class",
+			".$container_class.$container_class",
```
This change looks good to me. Is the reason we needed to do it because in the JS side of things we won't always have access to the block classname, is that right? Either way, I think the block classname was only there to increase specificity of this selector, so the change sounds good to me.

tellthemachines replied:
This is needed because the block class is attached to the outer wrapper, and because we're now adding the container class to the inner wrapper this style block will no longer work with that combination.

andrewserong reviewed (commented):
This is testing really well for me @tellthemachines, thanks for honing in on the final changes, it feels nice and clean and delivers on the promise of trying to get as close as possible to a "just works" approach for attaching layout classnames and styles to the inner blocks wrapper, and falling back to the outer wrapper where there is no inner blocks wrapper 👍

It feels like a good step toward a consistent API for stabilising the layout block support, too, so I think this would be a good approach to go with.

I've left a couple of questions on the PHP side of things to make sure I'm following along correctly 🙂

Also, I'd just like to ping @youknowriad if he has time to take a look, too, as I remember many earlier discussions in the development of the layout support, and mentions of one day moving to targeting the innerBlocks wrapper. So it might be good to confidence check that the approach is consistent with the earlier wish-list of how it might work 🙂.

This is very cool, though, I'm really excited about this one! 😀

tellthemachines committed:
Add comment to inner content classnames logic.

youknowriad commented on packages/block-editor/src/hooks/layout.js:
```diff
@@ -413,7 +408,7 @@ export const withLayoutStyles = createHigherOrderComponent(
 						element
 					) }
-				<BlockListBlock { ...props } className={ className } />
+				<BlockListBlock { ...props } layoutClassNames={ className } />
```
My original idea here was not to add a new prop to BlockListBlock and try to pass it down using context like we did but maybe we could try a dedicated filter for the inner blocks component instead of relying on `editor.BlockListBlock`.

Adding filters has its own drawbacks too so we must be careful regardless of the approach. Any approach requires a new API either a prop or a filter. I'd prefer an "internal/unstable" API.

tellthemachines replied:
If we make it a filter for InnerBlocks, we'll have to also create a hook to add the layout classnames to blocks that don't use InnerBlocks, like Gallery and the Placeholder version of Post Content. The advantage of passing the classnames through BlockListBlock and into BlockEdit is that we can retrieve them directly from the Edit props in those cases. 

I guess we can make this an "unstable" prop, but we do need to start thinking of what shape the layout API should take when stabilised. Even with this prop being unstable, the layout classnames will be output as props to the Edit function, which kind of makes them public by default 😅

youknowriad replied:
> the Placeholder version of Post Content

I think these get the "layout" from their attributes anyway and they already know that they support "layout"

tellthemachines replied:
They know their layout attributes, but we need to add the layout classnames manually because those blocks don't use InnerBlocks. We can't work out the classnames from the attributes because there's no way of knowing the `wp-container` class so there has to be a way to get the classnames in the blocks' edit function.

youknowriad reviewed (commented):
Nice work here

adamziel commented on lib/block-supports/layout.php:
```diff
@@ -412,16 +428,20 @@ function gutenberg_render_layout_support_flag( $block_content, $block ) {
+	$content = new WP_HTML_Tag_Processor( $block_content );
+	if ( $inner_content_classnames ) {
+		$content->next_tag( array( 'class_name' => $inner_content_classnames ) );
+		$content->add_class( implode( ' ', $class_names ) );
```
@tellthemachines This will work, but `add_class` is meant for adding one class at a time. Passing a long string with spaces won't give you the benefits of class deduplication, e.g.:

```php
$content = new WP_HTML_Tag_Processor('<div class="foo bar"></div>');
$content->next_tag();
$content->add_class('bar test');
echo $content.'';
// <div class="foo bar bar test"></div>
```

But:

```php
$content->add_class('bar');
$content->add_class('test');
echo $content.'';
// <div class="foo bar test"></div>
```

Now that I think about it, `WP_HTML_Tag_Processor` already prevents you from setting an attribute with an invalid name such as `foo =bar`. `add_class` should probably do the same.

It would also be convenient to accept arrays in `add_class` and `remove_class`, even if via splat like `->add_class('foo', 'bar');`. What do you think @dmsnell ?

dmsnell replied:
@adamziel easy enough to loop if given an array. I like `...$classes_to_add`; while not common in the WP codebase it's there in a couple of places. according to an overly-naïve benchmark passing an array is more efficient than varargs but insignificantly so - I believe we can prefer ergonomics here over performance.

to that end I feel like passing an array might be both more consistent and approachable and gives the opportunity for array re-use and so maybe that's the better option? probably overthinking it.

on the other hand it's easy to leave this complexity out of the class and let the caller do it and avoid complicating things that don't need to be.

```php
foreach ( strtok( $classes, ' ' ) as $classname ) {
	$tags->add_class( $classname );
}
```

and this sidelines all of the questions we created by proposing an array input. maybe leaving it be is the clearest option in lack of a clearly-dominant option. the abstraction isn't adding much and doesn't really make it easier to work with classes, since even if we add this additional calling convention, we still have to split the string and effectively come up with the same calling code, but duplicated logic in the class method.

so I think probably _no_, this is not a feature that would add much to the abstraction and should be left out.

tellthemachines replied:
I'll update this to loop through the array of classes and add each individually then 😄

ramonjd commented:
This is testing well for me! 

Each layout setting looks as it should in the editor and frontend:

https://user-images.githubusercontent.com/6458278/197441536-6f606339-1994-44fc-a4a5-3ed2c16d8ec3.mp4

The classnames are being applied to the inner block wrapper where appropriate. 

The constrained layout discrepancy between editor and frontend mentioned [above](https://github.com/WordPress/gutenberg/pull/44600#pullrequestreview-1139827718) appears to be fixed as well:

| Editor  | Frontend |
| ------------- | ------------- |
| <img width="400" alt="Screen Shot 2022-10-24 at 2 09 21 pm" src="https://user-images.githubusercontent.com/6458278/197441580-5663134a-bcb7-469b-b035-66313dafb8ab.png">  | <img width="400" alt="Screen Shot 2022-10-24 at 2 09 32 pm" src="https://user-images.githubusercontent.com/6458278/197441578-5292930a-ef3a-4297-a722-21be4204190b.png"> |

Checked Group, Column. Post Content and Buttons blocks and the layout functionality works as expected.


The only thing I noticed, which is not related to this PR, is that the layout controls appear a bit squished. Something for later. 

<img width="264" alt="Screen Shot 2022-10-24 at 2 17 36 pm" src="https://user-images.githubusercontent.com/6458278/197442465-60d40e53-11a1-42ff-bdd9-5842ae76effd.png">

tellthemachines committed:
Mark layoutClassnames unstable

tellthemachines committed:
Add classnames to container in a loop.

tellthemachines commented:
Thanks for the review @ramonjd ! I've addressed the remaining feedback:

* Marked `layoutClassnames` prop unstable;
* Looped through layout classnames to call `add_class` on each of them separately.

I'll now remove the Cover block layout support in preparation for merging. Additional testing can always be done by adding `"__experimentalLayout": true` to the Cover block.json locally.

tellthemachines committed:
Remove layout support for Cover block.

tellthemachines commented:
The e2e failures in the inserting blocks tests point to a legit issue with Group block in classic themes: the layout classnames are now being added to the inner instead of the outer wrapper. Which is of course the whole point of this PR 😅 but it breaks current styles applied to the classic Group block markup. Ideally we'd change the styles as these classes make more sense applied to the inner wrapper, but could that raise back compat issues for themes, if they're relying on the layout classname being applied to the outer wrapper?

tellthemachines committed:
Apply layout to outer wrapper if legacy Group

tellthemachines commented:
Update: I pushed a fix that should apply the layout classes on the outer wrapper for legacy Groups in the editor. Leaving the classes on the inner wrapper would break editor styling on Twenty Twenty so I thought it best to fix it this way. (In the front end classes are already being applied correctly)

ramonjd reviewed (approved):
I retested with the Cover block using the following diff:

<details>

<summary>Enable layout classes in Cover block:</summary>

```diff
diff --git a/packages/block-library/src/cover/block.json b/packages/block-library/src/cover/block.json
index 1982ecc448..f227ad7860 100644
--- a/packages/block-library/src/cover/block.json
+++ b/packages/block-library/src/cover/block.json
@@ -78,6 +78,12 @@
        },
        "usesContext": [ "postId", "postType" ],
        "supports": {
+               "__experimentalLayout": {
+                       "allowSwitching": true,
+                       "default": {
+                               "type": "constrained"
+                       }
+               },
                "anchor": true,
                "align": true,
                "html": false,
diff --git a/packages/block-library/src/cover/edit/index.js b/packages/block-library/src/cover/edit/index.js
index 4d7aba03dc..355a81250e 100644
--- a/packages/block-library/src/cover/edit/index.js
+++ b/packages/block-library/src/cover/edit/index.js
@@ -78,6 +78,7 @@ function CoverEdit( {
        setAttributes,
        setOverlayColor,
        toggleSelection,
+       __unstableLayoutClassNames: layoutClassNames,
        context: { postId, postType },
 } ) {
        const {
@@ -102,7 +103,7 @@ function CoverEdit( {
                'featured_media',
                postId
        );
-
+       console.log( 'layoutClassNames', layoutClassNames );
        const media = useSelect(
                ( select ) =>
                        featuredImage &&
@@ -185,7 +186,10 @@ function CoverEdit( {
 
        const innerBlocksProps = useInnerBlocksProps(
                {
-                       className: 'wp-block-cover__inner-container',
+                       className: classnames(
+                               'wp-block-cover__inner-container',
+                               layoutClassNames
+                       ),
                },
                {
                        template: innerBlocksTemplate,

```

</details>

The editor and frontend appear as expected. 👍

andrewserong commented on packages/block-editor/src/components/inner-blocks/index.js:
```diff
@@ -208,6 +211,7 @@ export function useInnerBlocksProps( props = {}, options = {} ) {
 			'block-editor-block-list__layout',
 			{
 				'has-overlay': hasOverlay,
+				[ layoutClassNames ]: __experimentalLayout,
```
It looks like this change might have broken the Columns block layout (the layout classnames don't appear to be output on the wrapper). For that block `__experimentalLayout` doesn't appear to be passed to explicitly down to `useInnerBlocksProps`. I wasn't sure if the current state is that it's required or not for it to be passed down, or if it's just an optional thing that blocks can pass down if they need to be explicit about the layout object being used?

<img width="834" alt="image" src="https://user-images.githubusercontent.com/14988353/197681609-4b7026a6-e2fc-4195-add9-21348ec2c203.png">

andrewserong replied:
It's possibly a slightly more rare use case this one: blocks that use layout but don't have any editable controls, so don't expect to pass down a `layout` object, and instead use the default layout object from `block.json`:

https://github.com/WordPress/gutenberg/blob/da9c992b418189c090ec0bd5055960e7f5680f01/packages/block-library/src/columns/block.json#L45-L48

andrewserong replied:
Apologies for all the separate comments here! I was curious about how it's hooked together (I couldn't quite remember). This appears to be the line where the fallback to the `block.json` version of the layout object is used?

https://github.com/WordPress/gutenberg/blob/008bb9c94ded2aacb1ea16478f7fb90c7d440215/packages/block-editor/src/hooks/layout.js#L361

Social links handles it here: https://github.com/WordPress/gutenberg/blob/a79329beb579b695349c5396ae66b6b342c4e178/packages/block-library/src/social-links/edit.js#L70

So maybe it should be fixed up in the Columns block? 🤔

tellthemachines replied:
> It looks like this change might have broken the Columns block layout (the layout classnames don't appear to be output on the wrapper)

Oh, good catch! I knew there was bound to be something wrong 😂 
Yeah it's a good question. I think with Columns it doesn't need layout passing into InnerBlocks because it only uses the default settings for the flex layout type, and doesn't show the controls so isn't user-adjustable.
Hmm I'll have a think about this. It's a bit awful but we need to _not_ add the layout classnames to the inner wrapper for theme back-compat reasons.

andrewserong replied:
Just an idea off the top of my head, but would it work to make `__unstableLayoutClassNames` overridable in `useInnerBlocksProps`? E.g. in the hook, use `__unstableLayoutClassNames` if it's passed in within `options`, and fallback to the context? Then in the Group block we could pass an empty string down to `useInnerBlocksProps` for the legacy classic theme case... or something like that?

tellthemachines replied:
> So maybe it should be fixed up in the Columns block?

I'm not sure. It'd be great to _not_ have to pass layout explicitly to innerBlocks in the edit function, and that's something I'd like to look at before we stabilise it.

andrewserong replied:
> It'd be great to not have to pass layout explicitly to innerBlocks in the edit function, and that's something I'd like to look at before we stabilise it.

Good thinking 👍

tellthemachines replied:
OOooh it should work if we just pass an empty string from Group. We shouldn't need any fallback because we just want those classes to _not_ be attached to the inner wrapper.

tellthemachines replied:
wait, that won't work because innerBlocks is grabbing the classnames directly from context 🤔

tellthemachines replied:
Yeah I think we'll have to go with something like you're saying @andrewserong! I'm going to make it an unstable prop that can be used to not apply layout classes at all. That way it's more like a skipSerialization type logic, where it won't output anything and the block is responsible for implementing whatever it needs to.

andrewserong replied:
That sounds good, and if the block really needs to output its own classnames and pass them down to inner props, it can do that via passing `props.className` anyway 👍

tellthemachines committed:
Off switch for layout classes in inner wrapper

andrewserong commented on packages/block-editor/src/components/inner-blocks/index.js:
```diff
@@ -150,7 +150,9 @@ const ForwardedInnerBlocks = forwardRef( ( props, ref ) => {
  */
 export function useInnerBlocksProps( props = {}, options = {} ) {
-	const { clientId } = useBlockEditContext();
+	const { __unstableNoLayoutClassNames } = options;
```
Tiny nit: for naming, I was wondering if `__unstableDisableLayoutClassNames` might give us a bit more consistency with some of the features (e.g. `disableLayoutStyles`), or are there more precedents for using `no` as the prefix?

tellthemachines replied:
Oooh that sounds much better, thanks for the suggestion!

andrewserong reviewed (approved):
This is testing great for me now @tellthemachines, thanks for all the follow-up here! Just left a question about the naming for the flag to pass down to `useInnerBlocksProps` but I don't feel strongly about it, so happy to go with whatever you think is best there.

I just gave this a smoke test with the following blocks (I _think_ this is all of the core blocks that currently opt-in to layout support from a quick scan of the `block.json` files):

✅ Layout support still works as expected for blocks: `buttons`, `column`, `columns`, `comments-pagination`, `gallery`, `group` (and row and stack variations), `navigation`, `post-content`, `post-template`, `query`, `query-pagination`, and `social-links`

LGTM! ✨

tellthemachines committed:
Change unstable prop name

ndiego commented:
Hi guys, I just discovered that this PR causes an issue for extenders. This [docs page](https://developer.wordpress.org/block-editor/reference-guides/filters/block-filters/#editor-blocklistblock) details how you can add custom classes to blocks. With this PR, any custom classes are added to both the main container as well as the inner block wrapper.

andrewserong commented:
Thanks for the heads-up @ndiego! I think I've found where this was happening and have opened up a proposed fix in #45499.
